### PR TITLE
ENH Kernel covariance 

### PR DESCRIPTION
**Pull request by kshedden**
This implements a simple method of moments estimator of a spatial covariance function for multivariate data, using the approach discussed on page 7 of this paper:

Genton M, W Kleiber (2015).  Cross covariance functions for multivariate geostatics.  Statistical Science 30(2). https://arxiv.org/pdf/1507.08017.pdf

### CODE REVIEW

**Comment by josef-pkt** [comment]
This pull request **introduces 1 alert** when merging dcc7e6d07e3ea52b820c12191d9695e20ca6f7e0 into 8928f8de2eb092fa249db40a5e6e4d42ceb54e99 - [view on LGTM.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-3dc54a9b29ae5f3aac98b657c789565c1c92d5c9)

**new alerts:**

* 1 for Signature mismatch in overriding method

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: statsmodels/stats/correlation_tools.py [hunk 1]
```diff
@@ -856,6 +857,176 @@ def corr_thresholded(data, minabs=None, max_elt=1e7):
...
+class GaussianMultivariateKernel(MultivariateKernel):
+    """
+    The Gaussian (squared exponential) multivariate kernel.
+    """
+
+    def call(self, x, loc):
+        return np.exp(-(x - loc)**2 / (2 * self.bw2)).sum(1) / self.bwk
+
+
+def kernel_covariance(exog, loc, group, kernel=None, bw=None):
```
Comment (josef-pkt) [inline statsmodels/stats/correlation_tools.py]: we use plural `groups`

**Comment by josef-pkt** [comment]
For now, I have mainly a question about the use case for this and related PRs:
Do you have the within group number of observations small enough to use a dense covariance matrix?
My impression from the outside was that many spatial applications have to move to using sparse matrices to handle the number of spatial or within group observations.

**Comment by kshedden** [comment]
Regarding sparse versus dense, this returns a function cov(x, y) that computes the cross-covariance matrix between Z(x) and Z(y), where Z(.) is a process on R^p (possibly with repeated independent observations).  The dimensions of d will generally be small (e.g. for human subject studies, height, weight, blood pressure at age x).  cov(x, y) in this case would return a 3x3 matrix, which is a covariance matrix if x=y and a cross-covariance matrix otherwise.  The underlying covariance function is infinite dimensional but we only ever evaluate it at one point.

This could be used for spatial data, but it is also useful for longitudinal or similarly-indexed data with many repeated observations, e.g. growth curves on individuals.  Since the observations can be made at arbitrary time points, and there can be different numbers of them per subject, it doesn't always work well with MixedLM and GEE.

The spatial application would normally not have repeated observations.  You could set `groups` to the same value for all rows of exog and this should still work, at least from a mechanical point of view.  But statistically, you would need fairly dense observations to get a useful estimate of the covariance. 

When there are replicates, any one subject can have just a few observations, because the underlying infinite-dimensional covariance function is shared by everyone, and we can pool data to gain power.

Comparing to ProcessRegression, this uses kernels rather than a parametric model for the covariance, and does not estimate the mean and covariance parameters jointly (here you would have to estimate the mean first and subtract it out).  The advantage of this approach is that it might be more flexible than the parametric approaches, and it can handle multivariate observations (e.g. joint model of temperature and humidity) rather than just one variable at a time.

**Comment by josef-pkt** [comment]
Thanks for the application.
So the basic structure is similar to GEE, except for the difference in estimators, and more "dimensions".

I was thinking of spatial panel data like what pysal handles, i.e. each `group` is a spatial cross section at a point or period in time, and we have many time periods.

**Comment by josef-pkt** [comment]
aside: I added topic-covtype as additional label
We are currently missing a cov_type for spatial correlation that uses a kernel for averaging. This sounds a bit similar. Currently, we have kernels in HAC but that is only for a single dimension, i.e. time. An extension that I thought about but didn't implement is to use the kernel on a multivariate dimension, e.g. applied to some spatial distance measure.
The idea is similar to cluster robust standard errors, but with kernel averaging when the clusters are too large to do unweighted averaging.  Also In one dimension like HAC, we currently assume equal spacing for the kernel averaging.)

**Comment by kshedden** [comment]
One use-case is imputation, taking values that are observed at arbitrary time points for different individuals, then imputing values on a common grid of time points for additional analyses.  Hence #4722 to make the covariance modeling and imputation work together a bit better.

**Comment by josef-pkt** [comment]
This pull request **introduces 2 alerts** when merging 0bfcee27b89b120e12290721b64430a4d8f049c2 into 8928f8de2eb092fa249db40a5e6e4d42ceb54e99 - [view on LGTM.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-14116ca7a9d99c76208d3299d7c3ccbbb69a2131)

**new alerts:**

* 1 for Signature mismatch in overriding method
* 1 for First argument of a method is not named &#39;self&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: statsmodels/stats/correlation_tools.py [hunk 2]
```diff
@@ -856,6 +863,176 @@ def corr_thresholded(data, minabs=None, max_elt=1e7):
...
+
+        for g, ii in ix.items():
+
+            m = len(ii)
+            j1, j2 = np.indices((m, m))
+            j1 = ii[j1.flat]
+            j2 = ii[j2.flat]
+            w = kx[j1] * ky[j2]
+
+            cm += np.einsum("ij,ik,i->jk", exog[j1, :], exog[j2, :], w)
```
Comment (josef-pkt) [inline statsmodels/stats/correlation_tools.py]: in GLM according to timing by @thequackdaddy  using something like the following was faster than einsum, but it was for a matrix version (GLM.hessian IIRC)
(exog[j1, :] * w).dot(exog[j2, :])`

**Comment by josef-pkt** [comment]
I think this is fine, and you can merge when ready.

I read through most of Genton, Kleiber to get a basic idea (but not more than that).

I think the current location in correlation_tools is fine because it returns a (nested) helper function instead of doing some full covariance computation.
We can move it if needed if or when we get some more general support for computing and using this kind of covariance and correlation matrices/functions. (I have no clear idea yet where to put all the correlation and covariance related function for general usage, stats, multivariate, robust, nonparametric, ...)

**Comment by josef-pkt** [comment]
This pull request **introduces 2 alerts** when merging 658ca7f36937eb2d6b1a60fe66cf53289b99f23b into 8928f8de2eb092fa249db40a5e6e4d42ceb54e99 - [view on LGTM.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-713814409a5a5bb12bb75120f42d3ae525b20d39)

**new alerts:**

* 1 for Signature mismatch in overriding method
* 1 for First argument of a method is not named &#39;self&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Pull request merged**